core/runtime/src/main/java/io/quarkus/runtime/StartupContext.java

```diff
@@ -21,21 +22,29 @@ public class StartupContext implements Closeable {
     // this is done to distinguish between the value having never been set and having been set as null
     private boolean lastValueSet = false;
     // the initial capacity was determined experimentally for a standard set of extensions
-    private final List<Runnable> shutdownTasks = new ArrayList<>(9);
-    private final List<Runnable> lastShutdownTasks = new ArrayList<>(7);
+    private final Deque<Runnable> shutdownTasks = new ConcurrentLinkedDeque<>();
+    private final Deque<Runnable> lastShutdownTasks = new ConcurrentLinkedDeque<>();
     private String[] commandLineArgs;
     private String currentBuildStepName;
 
     public StartupContext() {
         ShutdownContext shutdownContext = new ShutdownContext() {
             @Override
             public void addShutdownTask(Runnable runnable) {
-                shutdownTasks.add(runnable);
+                if (Objects.nonNull(runnable)) {
```

Comment:
Is the `Objects#nonNull()` really necessary here? I mean `runnable != null` is perfectly fine and even the javadoc of `Objects#nonNull()` states that the method exists to be used as a predicate/method reference.

Reply:
Hm, you're right - this kind of warning is useless. What if we change the code to `shutdownTasks.addFirst(Objects.requireNonNull(runnable));` and just throw an NPE? I think that it's reasonable to fail fast...

Reply:
Yes, in that case it could even make more sense to `throw new IllegalArgumentException()`...